cumulus/client/bootnodes/src/discovery.rs

```diff
@@ -0,0 +1,460 @@
+	/// terminated.
+	async fn maybe_start_discovery(&mut self) -> RelayChainResult<bool> {
+		// Start discovery if it is not currently in progress.
+		if self.key_being_discovered.is_none() &&
+			self.pending_responses.is_empty() &&
+			self.find_node_queries.is_empty()
+		{
+			// No need to start discovey again if the previous attempt succeeded.
+			if self.succeeded {
+				info!(
+					target: LOG_TARGET,
+					"Parachain bootnode discovery on the relay chain DHT succeeded",
+				);
+				Ok(false)
+			} else if self.attempts_left > 0 {
+				// No discovery in progress and we have attempts left, start discovery.
+				self.attempts_left -= 1;
+				self.start_discovery().await?;
+				Ok(true)
+			} else {
+				warn!(
+					target: LOG_TARGET,
+					"Failed to discover parachain bootnodes on the relay chain DHT after {} attempts, \
+					 giving up",
+					MAX_DISCOVERY_ATTEMPTS,
+				);
+				Ok(false)
+			}
+		} else {
+			// Discovery is already in progress, just continue the event loop.
+			Ok(true)
+		}
+	}
+
+	fn request_bootnode(&mut self, peer_id: PeerId) {
+		trace!(
+			target: LOG_TARGET,
+			"Requesting parachain bootnode from the relay chain {peer_id:?}",
+		);
+
+		let (tx, rx) = oneshot::channel();
+
+		self.relay_chain_network.start_request(
+			peer_id,
+			self.paranode_protocol_name.clone(),
+			self.para_id_scale_compact.clone(),
+			None,
+			tx,
+			IfDisconnected::TryConnect,
+		);
+
+		self.pending_responses.push(async move { (peer_id, rx.await) }.boxed());
+	}
+
+	fn handle_providers(&mut self, providers: Vec<PeerId>) {
+		debug!(
+			target: LOG_TARGET,
+			"Found parachain bootnode providers on the relay chain: {providers:?}",
+		);
+
+		for peer_id in providers {
+			if peer_id == self.relay_chain_network.local_peer_id() {
+				continue;
+			}
+
+			// libp2p may yield the same provider multiple times; skip if we alredy queried it.
+			if self.direct_requests.contains(&peer_id) || self.find_node_queries.contains(&peer_id)
+			{
+				continue;
+			}
+
+			// Directly request a bootnode from the peer without performing a `FIND_NODE` query
+			// first. With litep2p backend this will likely succeed, because cached provider
+			// addresses are automatically added to the transport manager known addresses list.
+			//
+			// With libp2p backend, or if the remote did not return the cached addresses of the
+			// provider, the request will fail and we will perform a `FIND_NODE` query.
+			self.direct_requests.insert(peer_id);
+			self.request_bootnode(peer_id);
+		}
+	}
+
+	fn handle_response(
+		&mut self,
+		peer_id: PeerId,
+		res: Result<Result<(Vec<u8>, ProtocolName), RequestFailure>, oneshot::Canceled>,
+	) {
+		let direct_request = self.direct_requests.remove(&peer_id);
+
+		let response = match res {
+			Ok(Ok((payload, _))) => match Response::decode(payload.as_slice()) {
+				Ok(response) => response,
+				Err(e) => {
+					warn!(
```

Comment:
At this point we could decrease the peer reputation? Since it provided data that does not comply w the protobuf encoding? 
This will make the current node eventually favour other peers